File: php/testcode/redirect_043.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_redirect_encrypted_session_target
function redirect043(BenchmarkRequest $req): BenchmarkResponse {
    $target = $_SESSION['redirect_after_login'] ?? '/dashboard';
    header('Location: ' . $target); // vuln-code-snippet safe-line php_redirect_encrypted_session_target
    return BenchmarkResponse::ok('');
}
// vuln-code-snippet end php_redirect_encrypted_session_target

File: rust/testcode/redirect_043.rs
//! CWE-601: Redirect target validated against server-side stored allowed return URL.

// vuln-code-snippet start testcodeRedirect043
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let next = req.param("next");
    let allowed_return = get_allowed_return_url();
    if next == allowed_return {
        let location = format!("Location: {}", next); // vuln-code-snippet target-line testcodeRedirect043
        super::shared::BenchmarkResponse::ok(&location)
    } else {
        super::shared::BenchmarkResponse::bad_request("Unexpected redirect")
    }
}

fn get_allowed_return_url() -> String {
    "/dashboard".to_string()
}
// vuln-code-snippet end testcodeRedirect043
